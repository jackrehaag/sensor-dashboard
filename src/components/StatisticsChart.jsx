import React from "react";
import Chart from "chart.js";

class StatisticsChart extends React.Component {
  state = {};
  chartRef = React.createRef();

  componentDidMount() {
    this.context = this.chartRef.current.getContext("2d");
    this.myChart = new Chart(this.context, {
      type: "line",
      data: {
        xAxisID: "time",
        yAxisID: "value",
        datasets: [
          {
            label: this.props.name,
            data: this.props.data,
            backgroundColor: ["rgba(97, 186, 255, 0.2)"],
            borderColor: ["rgba(97, 186, 255, 1)"],
            borderWidth: 1
          }
        ]
      },
      options: {
        elements: {
          line: {
            tension: 0 // disables bezier curves
          }
        },
        scales: {
          yAxes: [
            {
              ticks: {
                beginAtZero: true
              }
            }
          ],
          xAxes: [
            {
              id: "time",
              type: "time"
            }
          ]
        }
      }
    });
  }

  componentDidUpdate() {
    if (this.props.data) {
      this.myChart.data.datasets[0].data = this.props.data;
      this.myChart.update();
    }
  }

  render() {
    return (
      <div>
        <canvas ref={this.chartRef} width="400" height="270" />
      </div>
    );
  }
}

export default StatisticsChart;
