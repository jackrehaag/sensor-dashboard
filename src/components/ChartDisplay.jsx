import React from "react";
import ChartControls from "./ChartControls";
import StatisticsChart from "./StatisticsChart";
import getData from "../services/readings-api";

const TEMPERATURE_METRIC = "temperature";
const HUMIDITY_METRIC = "humidity";
const MOTION_DETECTION_METRIC = "motion_detection";

const date = new Date();
const defaultDate = date.toISOString().slice(0, 10);
const defaultMetric = TEMPERATURE_METRIC;

const metricDataSets = {
  [TEMPERATURE_METRIC]: "temperatureData",
  [HUMIDITY_METRIC]: "humidityData",
  [MOTION_DETECTION_METRIC]: "motionDetectionData"
};

class ChartDisplay extends React.Component {
  _isMounted = false;
  state = {
    activeData: [],
    dateSelection: defaultDate,
    metricSelection: defaultMetric
  };

  fetchNewData = async () => {
    let [
      temperatureData,
      humidityData,
      motionDetectionData
    ] = await Promise.all([
      getData(TEMPERATURE_METRIC, this.state.dateSelection),
      getData(HUMIDITY_METRIC, this.state.dateSelection),
      getData(MOTION_DETECTION_METRIC, this.state.dateSelection)
    ]);

    if (this._isMounted == true) {
      this.setState(
        {
          temperatureData: temperatureData,
          humidityData: humidityData,
          motionDetectionData: motionDetectionData
        },
        this.setActiveData
      );
    }
  };

  onDateChange = e => {
    e.preventDefault();
    this.setState({ dateSelection: e.target.value });
  };

  onDateSubmit = e => {
    e.preventDefault();
    this.fetchNewData();
  };

  componentDidMount() {
    this._isMounted = true;
    this.fetchNewData();
  }

  setActiveData = () => {
    if (this.state.metricSelection) {
      const dataSet = metricDataSets[this.state.metricSelection];
      if (dataSet && this.state[dataSet]) {
        this.setState({ activeData: this.state[dataSet] });
      }
    }
  };

  onMetricSelection = e => {
    e.preventDefault();
    this.setState({ metricSelection: e.target.value }, this.setActiveData);
  };

  componentWillUnmount() {
    this._isMounted = false;
  }

  render() {
    return (
      <div className="ui column">
        <ChartControls
          onDateChange={this.onDateChange}
          onSubmit={this.onDateSubmit}
          dateSelection={this.state.dateSelection}
          onMetricSelection={this.onMetricSelection}
        />
        <StatisticsChart
          name="Humidity percentage"
          data={this.state.activeData}
        />
      </div>
    );
  }
}

export default ChartDisplay;
