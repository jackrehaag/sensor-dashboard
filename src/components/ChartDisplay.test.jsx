import React from "react";
import ChartDisplay from "./ChartDisplay";
import renderer from "react-test-renderer";
jest.mock("chart.js");
jest.mock("./StatisticsChart");
import StatisticsChart from "./StatisticsChart";
jest.mock("../services/readings-api");
import Chart from "chart.js";
import Enzyme from "enzyme";
import getData from "../services/readings-api";

jest.mock("../services/readings-api", () => ({
  __esModule: true,
  default: jest.fn((metric, date) => {
    return new Promise((resolve, reject) => {
      if (metric == "temperature") {
        resolve([{ x: 1, y: 2 }]);
      } else if (metric == "humidity") {
        resolve([{ x: 3, y: 4 }]);
      } else if (metric == "motion_detection") {
        resolve([{ x: 5, y: 6 }]);
      } else {
        resolve([{ x: 0, y: 0 }]);
      }
    });
  })
}));

function createNodeMock(element) {
  if (element.type === "canvas") {
    return {
      getContext() {
        return {};
      }
    };
  }
}

test("it renders", () => {
  const component = renderer.create(<ChartDisplay />, { createNodeMock });
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

describe("interface functionality", () => {
  beforeEach(() => {
    getData.mockClear();
  });

  const component = Enzyme.mount(<ChartDisplay />, {
    disableLifestyleMethods: true
  });
  const temperatureButton = component.find('[value="temperature"]');
  const humidityButton = component.find('[value="humidity"]');
  const motionDetectionButton = component.find('[value="motion_detection"]');
  const dateInput = component.find("#chartControlDateField");
  const submitButton = component.find("button#chartControlSubmitButton");

  test("it updates the state when the temperature button is clicked", () => {
    temperatureButton.simulate("click");
    expect(component.state()).toEqual({
      activeData: [{ x: 1, y: 2 }],
      dateSelection: "2019-07-02",
      metricSelection: "temperature",
      temperatureData: [{ x: 1, y: 2 }],
      humidityData: [{ x: 3, y: 4 }],
      motionDetectionData: [{ x: 5, y: 6 }]
    });
  });

  test("it updates the state when the humidity button is clicked", () => {
    humidityButton.simulate("click");
    expect(component.state()).toEqual({
      activeData: [{ x: 3, y: 4 }],
      dateSelection: "2019-07-02",
      metricSelection: "humidity",
      temperatureData: [{ x: 1, y: 2 }],
      humidityData: [{ x: 3, y: 4 }],
      motionDetectionData: [{ x: 5, y: 6 }]
    });
  });

  test("it updates the state when the motion detection button is clicked", () => {
    motionDetectionButton.simulate("click");
    expect(component.state()).toEqual({
      activeData: [{ x: 5, y: 6 }],
      dateSelection: "2019-07-02",
      metricSelection: "motion_detection",
      temperatureData: [{ x: 1, y: 2 }],
      humidityData: [{ x: 3, y: 4 }],
      motionDetectionData: [{ x: 5, y: 6 }]
    });
  });

  test("it updates the state when the date is changed", () => {
    dateInput.simulate("change", { target: { value: "2019-07-03" } });
    expect(component.state()).toEqual({
      activeData: [{ x: 5, y: 6 }],
      dateSelection: "2019-07-03",
      metricSelection: "motion_detection",
      temperatureData: [{ x: 1, y: 2 }],
      humidityData: [{ x: 3, y: 4 }],
      motionDetectionData: [{ x: 5, y: 6 }]
    });
  });

  test("it sends a request when the submit button is pressed", () => {
    submitButton.simulate("submit");
    expect(getData.mock.calls).toEqual([
      ["temperature", "2019-07-03"],
      ["humidity", "2019-07-03"],
      ["motion_detection", "2019-07-03"]
    ]);
  });

  test("it sets the correct data set as activeData", () => {
    humidityButton.simulate("click");
    component.update();
    submitButton.simulate("submit");

    expect(component.state()).toEqual({
      activeData: [{ x: 3, y: 4 }],
      dateSelection: "2019-07-03",
      metricSelection: "humidity",
      temperatureData: [{ x: 1, y: 2 }],
      humidityData: [{ x: 3, y: 4 }],
      motionDetectionData: [{ x: 5, y: 6 }]
    });
  });
});
