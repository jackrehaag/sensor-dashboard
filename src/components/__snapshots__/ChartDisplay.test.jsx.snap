// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
<div
  className="ui column"
>
  <div>
    <button
      className="ui button small"
      id="chartControlTemperatureButton"
      onClick={[Function]}
      type="button"
      value="temperature"
    >
      <i
        className="thermometer icon"
      />
      Temperature
    </button>
    <button
      className="ui button small"
      id="chartControlHumidityButton"
      onClick={[Function]}
      type="button"
      value="humidity"
    >
      <i
        className="tint icon"
      />
      Humidity
    </button>
    <button
      className="ui button small"
      id="chartControlMotionDetectionButton"
      onClick={[Function]}
      type="button"
      value="motion_detection"
    >
      <i
        className="street view icon"
      />
      Motion detection
    </button>
    <form
      className="ui form"
      onSubmit={[Function]}
      style={
        Object {
          "display": "inline",
          "padding": ".5833em .833em",
        }
      }
    >
      <div
        className="ui action small input"
      >
        <input
          id="chartControlDateField"
          onChange={[Function]}
          placeholder="Date"
          type="text"
          value="2019-07-02"
        />
        <button
          className="ui button"
          id="chartControlSubmitButton"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;
